File: extensions/pi-looper/automations/cleanup-completed-worker-worktrees.py
#!/usr/bin/env python3
"""Deterministically clean completed pi-looper Herdr worker worktrees.

The script has two modes:

- plan mode (default): inspect GitHub PRs, Herdr worktrees, and git status, then
  report removable worker worktrees.
- apply mode: execute the plan without force-removing dirty worktrees.

It intentionally keeps the removal decision outside LLM prompts so the issue
coordinator only has to invoke this script and report its result.
"""

from __future__ import annotations

import argparse
import json
import os
import subprocess
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Any


DEFAULT_REVIEW_LABEL = "agent:review"
DEFAULT_HUMAN_LABEL = "ready-for-human"


@dataclass(frozen=True)
class Config:
    repo: str
    repo_path: str
    worktree_root: str
    review_label: str
    human_label: str


class CommandError(RuntimeError):
    pass


class CommandRunner:
    def json(self, args: list[str], cwd: str | None = None) -> Any:
        output = self.text(args, cwd=cwd)
        return json.loads(output)

    def text(self, args: list[str], cwd: str | None = None, check: bool = True) -> str:
        try:
            result = subprocess.run(args, cwd=cwd, text=True, capture_output=True, check=False)
        except FileNotFoundError as error:
            if check:
                raise CommandError(str(error)) from error
            return ""
        if check and result.returncode != 0:
            raise CommandError((result.stderr or result.stdout or f"command failed: {args!r}").strip())
        return result.stdout

    def code(self, args: list[str], cwd: str | None = None) -> int:
        try:
            return subprocess.run(args, cwd=cwd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, check=False).returncode
        except FileNotFoundError:
            return 127


def labels_of(item: dict[str, Any]) -> set[str]:
    labels = item.get("labels") or []
    names: set[str] = set()
    for label in labels:
        if isinstance(label, str):
            names.add(label)
        elif isinstance(label, dict) and label.get("name"):
            names.add(str(label["name"]))
    return names


def is_merged_pr(pr: dict[str, Any]) -> bool:
    return str(pr.get("state") or "").upper() == "MERGED" or bool(pr.get("mergedAt"))


def is_closed_pr(pr: dict[str, Any]) -> bool:
    state = str(pr.get("state") or "").upper()
    return state in {"CLOSED", "MERGED"} or bool(pr.get("closedAt")) or bool(pr.get("mergedAt"))


def is_pi_looper_pr(pr: dict[str, Any], config: Config) -> bool:
    branch = str(pr.get("headRefName") or "")
    return branch.startswith("agent/issue-") or bool(labels_of(pr) & {config.review_label, config.human_label})


def norm_path(value: str) -> str:
    return os.path.abspath(os.path.expanduser(value))


def is_under_root(path: str, root: str) -> bool:
    if not root:
        return True
    try:
        return os.path.commonpath([norm_path(path), norm_path(root)]) == norm_path(root)
    except ValueError:
        return False


def is_clean_status(status: Any) -> bool:
    return str(status or "").strip() == ""


def local_head_matches_closed_pr(worktree: dict[str, Any], pr: dict[str, Any], git_heads: dict[str, str], runner: CommandRunner | None) -> bool:
    head_oid = str(pr.get("headRefOid") or "")
    if not head_oid:
        return False
    path = str(worktree.get("path") or "")
    local_head = git_heads.get(path)
    if local_head is None and runner is not None:
        local_head = runner.text(["git", "-C", path, "rev-parse", "HEAD"], check=False).strip()
    if not local_head:
        return False
    if local_head == head_oid:
        return True
    if runner is None:
        return False
    return runner.code(["git", "-C", path, "merge-base", "--is-ancestor", local_head, head_oid]) == 0


def skip(reason: str, pr: dict[str, Any], worktree: dict[str, Any]) -> dict[str, Any]:
    return {
        "reason": reason,
        "prNumber": pr.get("number"),
        "branch": worktree.get("branch") or pr.get("headRefName"),
        "path": worktree.get("path"),
        "workspaceId": worktree.get("open_workspace_id"),
    }


def candidate(pr: dict[str, Any], worktree: dict[str, Any], reason: str) -> dict[str, Any]:
    return {
        "prNumber": pr.get("number"),
        "branch": worktree.get("branch"),
        "path": worktree.get("path"),
        "workspaceId": worktree.get("open_workspace_id"),
        "reason": reason,
    }


def select_cleanup_plan(
    *,
    config: Config,
    prs: list[dict[str, Any]],
    worktrees: list[dict[str, Any]],
    git_statuses: dict[str, str],
    git_heads: dict[str, str] | None = None,
    runner: CommandRunner | None = None,
) -> dict[str, list[dict[str, Any]]]:
    git_heads = git_heads or {}
    by_branch = {str(worktree.get("branch") or ""): worktree for worktree in worktrees if worktree.get("branch")}
    candidates: list[dict[str, Any]] = []
    skipped: list[dict[str, Any]] = []
    selected_paths: set[str] = set()

    for pr in sorted(prs, key=lambda item: int(item.get("number") or 0)):
        branch = str(pr.get("headRefName") or "")
        if not branch or not is_closed_pr(pr) or not is_pi_looper_pr(pr, config):
            continue
        worktree = by_branch.get(branch)
        if not worktree:
            continue

        path = str(worktree.get("path") or "")
        if not path:
            skipped.append(skip("missing_path", pr, worktree))
            continue
        if norm_path(path) == norm_path(config.repo_path):
            skipped.append(skip("main_workspace", pr, worktree))
            continue
        if worktree.get("is_linked_worktree") is False:
            skipped.append(skip("not_linked_worktree", pr, worktree))
            continue
        if not worktree.get("open_workspace_id"):
            skipped.append(skip("missing_workspace_id", pr, worktree))
            continue
        if not is_under_root(path, config.worktree_root):
            skipped.append(skip("outside_worktree_root", pr, worktree))
            continue

        status = git_statuses.get(path)
        if status is None and runner is not None:
            try:
                status = runner.text(["git", "-C", path, "status", "--short"])
            except CommandError:
                skipped.append(skip("status_unavailable", pr, worktree))
                continue
        if status is None:
            skipped.append(skip("status_unavailable", pr, worktree))
            continue
        if not is_clean_status(status):
            skipped.append(skip("dirty_worktree", pr, worktree))
            continue

        if is_merged_pr(pr):
            reason = "merged_pr"
        elif local_head_matches_closed_pr(worktree, pr, git_heads, runner):
            reason = "closed_pr_head_preserved"
        else:
            skipped.append(skip("closed_pr_head_not_verified", pr, worktree))
            continue

        if path in selected_paths:
            continue
        selected_paths.add(path)
        candidates.append(candidate(pr, worktree, reason))

    return {"candidates": candidates, "skipped": skipped}


def config_from_env() -> Config:
    return Config(
        repo=os.environ["PI_LOOPER_GITHUB_REPO"],
        repo_path=os.environ["PI_LOOPER_REPO_PATH"],
        worktree_root=os.environ.get("PI_LOOPER_WORKTREE_ROOT", ""),
        review_label=os.environ.get("PI_LOOPER_REVIEW_LABEL", DEFAULT_REVIEW_LABEL),
        human_label=os.environ.get("PI_LOOPER_HUMAN_LABEL", DEFAULT_HUMAN_LABEL),
    )


def config_from_fixture(data: dict[str, Any]) -> Config:
    return Config(
        repo=str(data.get("repo") or "owner/repo"),
        repo_path=str(data.get("repoPath") or "/repo"),
        worktree_root=str(data.get("worktreeRoot") or ""),
        review_label=str(data.get("reviewLabel") or DEFAULT_REVIEW_LABEL),
        human_label=str(data.get("humanLabel") or DEFAULT_HUMAN_LABEL),
    )


def load_live_plan(config: Config, runner: CommandRunner) -> dict[str, list[dict[str, Any]]]:
    prs: list[dict[str, Any]] = []
    seen: set[int] = set()
    for state in ("merged", "closed"):
        try:
            batch = runner.json([
                "gh",
                "pr",
                "list",
                "-R",
                config.repo,
                "--state",
                state,
                "--limit",
                "100",
                "--json",
                "number,state,mergedAt,closedAt,headRefName,headRefOid,labels",
            ])
        except (CommandError, json.JSONDecodeError):
            batch = []
        for pr in batch:
            number = int(pr.get("number") or 0)
            if number in seen:
                continue
            seen.add(number)
            prs.append(pr)

    worktree_data = runner.json(["herdr", "worktree", "list", "--cwd", config.repo_path, "--json"])
    worktrees = ((worktree_data.get("result") or {}).get("worktrees") or []) if isinstance(worktree_data, dict) else []
    return select_cleanup_plan(config=config, prs=prs, worktrees=worktrees, git_statuses={}, runner=runner)


def load_fixture_plan(path: str) -> dict[str, list[dict[str, Any]]]:
    data = json.loads(Path(path).read_text())
    return select_cleanup_plan(
        config=config_from_fixture(data),
        prs=data.get("prs") or [],
        worktrees=data.get("worktrees") or [],
        git_statuses=(data.get("git") or {}).get("statuses") or {},
        git_heads=(data.get("git") or {}).get("heads") or {},
        runner=None,
    )


def apply_plan(plan: dict[str, list[dict[str, Any]]], config: Config, runner: CommandRunner) -> dict[str, Any]:
    runner.text(["git", "-C", config.repo_path, "fetch", "--prune"], check=False)
    removed: list[dict[str, Any]] = []
    failed: list[dict[str, Any]] = []

    for item in plan["candidates"]:
        workspace_id = item.get("workspaceId")
        try:
            if not workspace_id:
                raise CommandError("missing Herdr workspace id; refusing direct git worktree removal")
            runner.text(["herdr", "worktree", "remove", "--workspace", str(workspace_id), "--json"])
            removed.append(item)
        except CommandError as error:
            failed.append({**item, "error": str(error)})

    return {**plan, "removed": removed, "failed": failed}


def parse_args(argv: list[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--apply", action="store_true", help="remove selected worktrees")
    parser.add_argument("--plan", action="store_true", help="only print the cleanup plan (default)")
    parser.add_argument("--json", action="store_true", help="print JSON output")
    parser.add_argument("--fixture", help="load all inputs from a fixture JSON file")
    return parser.parse_args(argv)


def main(argv: list[str]) -> int:
    args = parse_args(argv)
    runner = CommandRunner()

    if args.fixture:
        plan = load_fixture_plan(args.fixture)
        result: dict[str, Any] = plan
    else:
        config = config_from_env()
        plan = load_live_plan(config, runner)
        result = apply_plan(plan, config, runner) if args.apply else plan

    if args.json:
        print(json.dumps(result, ensure_ascii=False, sort_keys=True))
    else:
        print(f"cleanup candidates: {len(result.get('candidates') or [])}")
        if result.get("removed") is not None:
            print(f"removed: {len(result.get('removed') or [])}")
        if result.get("failed"):
            print(f"failed: {len(result.get('failed') or [])}")
    return 1 if result.get("failed") else 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))

File: extensions/deadloop/automations/cleanup-completed-worker-worktrees.ts
#!/usr/bin/env node
// Deterministically clean completed deadloop Herdr worker worktrees.
// CommonJS-shaped so it can run directly with `node cleanup-completed-worker-worktrees.ts`.

const fs = require("node:fs") as typeof import("node:fs");
const os = require("node:os") as typeof import("node:os");
const path = require("node:path") as typeof import("node:path");
const { spawnSync } = require("node:child_process") as typeof import("node:child_process");
const { createHerdrRunner, normalizeHerdrWorktreeRecord } = require("../../../src/herdr-runner.ts");
const { withEnabledDriverLock } = require("../../../src/driver-enablement.cjs");

type CleanupRecord = Record<string, any>;

type CleanupConfig = {
  repo: string;
  repoPath: string;
  worktreeRoot: string;
  reviewLabel: string;
  humanLabel: string;
  stateDir?: string;
  enabledAt?: number;
};

const DEFAULT_REVIEW_LABEL = "agent:review";
const DEFAULT_HUMAN_LABEL = "ready-for-human";

function runCleanupText(args: string[], options: { cwd?: string; check?: boolean } = {}): string {
  const result = spawnSync(args[0], args.slice(1), {
    cwd: options.cwd,
    encoding: "utf8",
    stdio: ["ignore", "pipe", "pipe"],
  });
  if (options.check !== false && result.status !== 0) {
    throw new Error((result.stderr || result.stdout || `command failed: ${args.join(" ")}`).trim());
  }
  return result.stdout || "";
}

function runCleanupJson(args: string[], options: { cwd?: string } = {}): any {
  return JSON.parse(runCleanupText(args, options));
}

function cleanupHerdrRunner() {
  return createHerdrRunner({
    runText: (command: string, args: string[]) => runCleanupText([command, ...args]),
    runJson: (command: string, args: string[]) => runCleanupJson([command, ...args]),
  });
}

function workspaceIdForCleanup(worktree: CleanupRecord): string {
  return String(worktree.workspaceId || "");
}

function runCleanupCode(args: string[], options: { cwd?: string } = {}): number {
  const result = spawnSync(args[0], args.slice(1), {
    cwd: options.cwd,
    encoding: "utf8",
    stdio: ["ignore", "ignore", "ignore"],
  });
  return result.status ?? 127;
}

function labelsOfCleanup(item: CleanupRecord): Set<string> {
  const names = new Set<string>();
  for (const label of item.labels || []) {
    if (typeof label === "string") names.add(label);
    else if (label && typeof label === "object" && label.name) names.add(String(label.name));
  }
  return names;
}

function isMergedPrForCleanup(pr: CleanupRecord): boolean {
  return String(pr.state || "").toUpperCase() === "MERGED" || Boolean(pr.mergedAt);
}

function isClosedPrForCleanup(pr: CleanupRecord): boolean {
  const state = String(pr.state || "").toUpperCase();
  return ["CLOSED", "MERGED"].includes(state) || Boolean(pr.closedAt) || Boolean(pr.mergedAt);
}

function isPiLooperPrForCleanup(pr: CleanupRecord, config: CleanupConfig): boolean {
  const branch = String(pr.headRefName || "");
  return branch.startsWith("agent/issue-") || [config.reviewLabel, config.humanLabel].some((label) => labelsOfCleanup(pr).has(label));
}

function expandHomeForCleanup(value: string): string {
  return value === "~" || value.startsWith("~/") ? path.join(os.homedir(), value.slice(2)) : value;
}

function normPathForCleanup(value: string): string {
  return path.resolve(expandHomeForCleanup(value));
}

function isUnderRootForCleanup(candidatePath: string, root: string): boolean {
  if (!root) return true;
  const absolutePath = normPathForCleanup(candidatePath);
  const absoluteRoot = normPathForCleanup(root);
  const relative = path.relative(absoluteRoot, absolutePath);
  return relative === "" || (!relative.startsWith("..") && !path.isAbsolute(relative));
}

function isGeneratedAgentArtifactStatusLine(line: string): boolean {
  return /^\?\?\s+\.(?:deadloop|pi-subagents)(?:\/|$)/.test(line.trim());
}

function isCleanStatusForCleanup(status: unknown): boolean {
  return String(status || "")
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter(Boolean)
    .every(isGeneratedAgentArtifactStatusLine);
}

function localHeadMatchesClosedPr(
  worktree: CleanupRecord,
  pr: CleanupRecord,
  gitHeads: Record<string, string>,
  live: boolean,
): boolean {
  const headOid = String(pr.headRefOid || "");
  if (!headOid) return false;
  const worktreePath = String(worktree.path || "");
  let localHead = gitHeads[worktreePath];
  if (localHead === undefined && live) {
    localHead = runCleanupText(["git", "-C", worktreePath, "rev-parse", "HEAD"], { check: false }).trim();
  }
  if (!localHead) return false;
  if (localHead === headOid) return true;
  if (!live) return false;
  return runCleanupCode(["git", "-C", worktreePath, "merge-base", "--is-ancestor", localHead, headOid]) === 0;
}

function skipCleanup(reason: string, pr: CleanupRecord, worktree: CleanupRecord): CleanupRecord {
  return {
    reason,
    prNumber: pr.number,
    branch: worktree.branch || pr.headRefName,
    path: worktree.path,
    workspaceId: workspaceIdForCleanup(worktree),
  };
}

function cleanupCandidate(pr: CleanupRecord, worktree: CleanupRecord, reason: string): CleanupRecord {
  return {
    prNumber: pr.number,
    branch: worktree.branch,
    path: worktree.path,
    workspaceId: workspaceIdForCleanup(worktree),
    reason,
  };
}

function selectCleanupPlan({
  config,
  prs,
  worktrees,
  gitStatuses,
  gitHeads = {},
  live = false,
}: {
  config: CleanupConfig;
  prs: CleanupRecord[];
  worktrees: CleanupRecord[];
  gitStatuses: Record<string, string>;
  gitHeads?: Record<string, string>;
  live?: boolean;
}): { candidates: CleanupRecord[]; skipped: CleanupRecord[] } {
  const byBranch = new Map<string, CleanupRecord>();
  for (const worktree of worktrees) if (worktree.branch) byBranch.set(String(worktree.branch), worktree);
  const candidates: CleanupRecord[] = [];
  const skipped: CleanupRecord[] = [];
  const selectedPaths = new Set<string>();

  for (const pr of [...prs].sort((left, right) => Number(left.number || 0) - Number(right.number || 0))) {
    const branch = String(pr.headRefName || "");
    if (!branch || !isClosedPrForCleanup(pr) || !isPiLooperPrForCleanup(pr, config)) continue;
    const worktree = byBranch.get(branch);
    if (!worktree) continue;

    const worktreePath = String(worktree.path || "");
    if (!worktreePath) {
      skipped.push(skipCleanup("missing_path", pr, worktree));
      continue;
    }
    if (normPathForCleanup(worktreePath) === normPathForCleanup(config.repoPath)) {
      skipped.push(skipCleanup("main_workspace", pr, worktree));
      continue;
    }
    if (worktree.is_linked_worktree === false) {
      skipped.push(skipCleanup("not_linked_worktree", pr, worktree));
      continue;
    }
    if (!workspaceIdForCleanup(worktree)) {
      skipped.push(skipCleanup("missing_workspace_id", pr, worktree));
      continue;
    }
    if (!isUnderRootForCleanup(worktreePath, config.worktreeRoot)) {
      skipped.push(skipCleanup("outside_worktree_root", pr, worktree));
      continue;
    }

    let status = gitStatuses[worktreePath];
    if (status === undefined && live) {
      try {
        status = runCleanupText(["git", "-C", worktreePath, "status", "--short"]);
      } catch {
        skipped.push(skipCleanup("status_unavailable", pr, worktree));
        continue;
      }
    }
    if (status === undefined) {
      skipped.push(skipCleanup("status_unavailable", pr, worktree));
      continue;
    }
    if (!isCleanStatusForCleanup(status)) {
      skipped.push(skipCleanup("dirty_worktree", pr, worktree));
      continue;
    }

    let reason: string;
    if (isMergedPrForCleanup(pr)) {
      reason = "merged_pr";
    } else if (localHeadMatchesClosedPr(worktree, pr, gitHeads, live)) {
      reason = "closed_pr_head_preserved";
    } else {
      skipped.push(skipCleanup("closed_pr_head_not_verified", pr, worktree));
      continue;
    }

    if (selectedPaths.has(worktreePath)) continue;
    selectedPaths.add(worktreePath);
    candidates.push(cleanupCandidate(pr, worktree, reason));
  }

  return { candidates, skipped };
}

function cleanupConfigFromEnv(): CleanupConfig {
  return {
    repo: process.env.DEADLOOP_GITHUB_REPO || "",
    repoPath: process.env.DEADLOOP_REPO_PATH || "",
    worktreeRoot: process.env.DEADLOOP_WORKTREE_ROOT || "",
    reviewLabel: process.env.DEADLOOP_REVIEW_LABEL || DEFAULT_REVIEW_LABEL,
    humanLabel: process.env.DEADLOOP_HUMAN_LABEL || DEFAULT_HUMAN_LABEL,
    stateDir: process.env.DEADLOOP_STATE_DIR,
    enabledAt: process.env.DEADLOOP_ENABLED_AT === undefined ? undefined : Number(process.env.DEADLOOP_ENABLED_AT),
  };
}

function cleanupConfigFromFixture(data: CleanupRecord): CleanupConfig {
  return {
    repo: String(data.repo || "owner/repo"),
    repoPath: String(data.repoPath || "/repo"),
    worktreeRoot: String(data.worktreeRoot || ""),
    reviewLabel: String(data.reviewLabel || DEFAULT_REVIEW_LABEL),
    humanLabel: String(data.humanLabel || DEFAULT_HUMAN_LABEL),
  };
}

function loadLiveCleanupPlan(config: CleanupConfig): { candidates: CleanupRecord[]; skipped: CleanupRecord[] } {
  const prs: CleanupRecord[] = [];
  const seen = new Set<number>();
  for (const state of ["merged", "closed"]) {
    let batch: CleanupRecord[] = [];
    try {
      batch = runCleanupJson([
        "gh",
        "pr",
        "list",
        "-R",
        config.repo,
        "--state",
        state,
        "--limit",
        "100",
        "--json",
        "number,state,mergedAt,closedAt,headRefName,headRefOid,labels",
      ]);
    } catch {
      batch = [];
    }
    for (const pr of batch) {
      const number = Number(pr.number || 0);
      if (seen.has(number)) continue;
      seen.add(number);
      prs.push(pr);
    }
  }

  const worktrees = cleanupHerdrRunner().listWorktrees(config.repoPath);
  return selectCleanupPlan({ config, prs, worktrees, gitStatuses: {}, live: true });
}

function loadFixtureCleanupPlan(file: string): { candidates: CleanupRecord[]; skipped: CleanupRecord[] } {
  const data = JSON.parse(fs.readFileSync(file, "utf8"));
  return selectCleanupPlan({
    config: cleanupConfigFromFixture(data),
    prs: data.prs || [],
    worktrees: (data.worktrees || []).map(normalizeHerdrWorktreeRecord),
    gitStatuses: data.git?.statuses || {},
    gitHeads: data.git?.heads || {},
  });
}

function withCleanupMutation<T>(config: CleanupConfig, mutation: () => T): T {
  if (config.enabledAt === undefined) return mutation();
  if (!Number.isFinite(config.enabledAt) || !config.stateDir) {
    throw new Error("cleanup enablement generation is invalid");
  }
  return withEnabledDriverLock({
    repoPath: config.repoPath,
    githubRepo: config.repo,
    stateDir: config.stateDir,
    enabledAt: config.enabledAt,
  }, (_enabled: unknown, recheck: () => void) => {
    recheck();
    return mutation();
  });
}

function removeGeneratedAgentArtifacts(worktreePath: string, config: CleanupConfig): void {
  const tracked = runCleanupText(["git", "-C", worktreePath, "ls-files", "-z", "--", ".deadloop", ".pi-subagents"])
    .split("\0")
    .filter(Boolean);
  if (tracked.length) throw new Error(`runtime-named directories contain tracked files; refusing cleanup: ${tracked.join(", ")}`);

  const currentStatus = runCleanupText(["git", "-C", worktreePath, "status", "--short"]);
  if (!isCleanStatusForCleanup(currentStatus)) throw new Error("worktree became dirty after cleanup planning; refusing removal");

  for (const directory of [".deadloop", ".pi-subagents"]) {
    withCleanupMutation(config, () => fs.rmSync(path.join(worktreePath, directory), { recursive: true, force: true }));
  }
}

function applyCleanupPlan(plan: { candidates: CleanupRecord[]; skipped: CleanupRecord[] }, config: CleanupConfig): CleanupRecord {
  runCleanupText(["git", "-C", config.repoPath, "fetch", "--prune"], { check: false });
  const removed: CleanupRecord[] = [];
  const failed: CleanupRecord[] = [];

  for (const item of plan.candidates) {
    const workspaceId = item.workspaceId;
    try {
      if (!workspaceId) throw new Error("missing Herdr workspace id; refusing direct git worktree removal");
      const worktreePath = String(item.path || "");
      if (!worktreePath) throw new Error("missing worktree path; refusing cleanup");
      removeGeneratedAgentArtifacts(worktreePath, config);
      const remainingStatus = runCleanupText(["git", "-C", worktreePath, "status", "--short"]);
      if (remainingStatus.trim()) throw new Error("worktree became dirty after cleanup planning; refusing removal");
      withCleanupMutation(config, () => cleanupHerdrRunner().removeWorktree(String(workspaceId)));
      removed.push(item);
    } catch (error) {
      failed.push({ ...item, error: error instanceof Error ? error.message : String(error) });
    }
  }

  return { ...plan, removed, failed };
}

function requiredCleanupValue(argv: string[], index: number, flag: string): string {
  const value = argv[index + 1];
  if (!value || value.startsWith("--")) throw new Error(`${flag} requires a value`);
  return value;
}

function parseCleanupArgs(argv: string[]): CleanupRecord {
  const parsed: CleanupRecord = { apply: false, plan: false, json: false };
  for (let index = 0; index < argv.length; index += 1) {
    const token = argv[index];
    if (token === "--apply") {
      parsed.apply = true;
      continue;
    }
    if (token === "--plan") {
      parsed.plan = true;
      continue;
    }
    if (token === "--json") {
      parsed.json = true;
      continue;
    }
    if (token === "--fixture") {
      parsed.fixture = requiredCleanupValue(argv, index, token);
      index += 1;
      continue;
    }
    if (token.startsWith("--fixture=")) {
      parsed.fixture = token.slice("--fixture=".length);
      continue;
    }
    if (token === "--help" || token === "-h") {
      parsed.help = true;
      continue;
    }
    throw new Error(`unknown flag: ${token}`);
  }
  return parsed;
}

function cleanupHelp(): string {
  return "Usage: cleanup-completed-worker-worktrees.ts [--plan|--apply] [--json] [--fixture FILE]";
}

function main(argv: string[] = process.argv.slice(2)): number {
  const args = parseCleanupArgs(argv);
  if (args.help) {
    process.stdout.write(`${cleanupHelp()}\n`);
    return 0;
  }

  let result: CleanupRecord;
  if (args.fixture) {
    result = loadFixtureCleanupPlan(args.fixture);
  } else {
    const config = cleanupConfigFromEnv();
    if (!config.repo || !config.repoPath) throw new Error("DEADLOOP_GITHUB_REPO and DEADLOOP_REPO_PATH are required");
    const plan = loadLiveCleanupPlan(config);
    result = args.apply ? applyCleanupPlan(plan, config) : plan;
  }

  if (args.json) {
    process.stdout.write(`${JSON.stringify(result)}\n`);
  } else {
    process.stdout.write(`cleanup candidates: ${(result.candidates || []).length}\n`);
    if (result.removed !== undefined) process.stdout.write(`removed: ${(result.removed || []).length}\n`);
    if (result.failed?.length) process.stdout.write(`failed: ${result.failed.length}\n`);
  }
  return result.failed?.length ? 1 : 0;
}

if (require.main === module) {
  try {
    process.exitCode = main();
  } catch (error) {
    console.error(`cleanup-completed-worker-worktrees.ts: ${error instanceof Error ? error.message : String(error)}`);
    process.exitCode = 2;
  }
}

module.exports = {
  applyCleanupPlan,
  loadFixtureCleanupPlan,
  selectCleanupPlan,
};
